src/app/login-form/login-form-register.component.ts

```diff
@@ -0,0 +1,65 @@
+import {Component, Inject} from '@angular/core';
+import {MAT_DIALOG_DATA, MatDialogRef} from '@angular/material';
+import {User, UserService} from '../api/user.service';
+
+import {FormControl, Validators} from '@angular/forms';
+import { MatSnackBar } from '@angular/material';
+
+@Component({
+  selector: `app-login-form-register.component`,
+  templateUrl: 'login-form-register.component.html',
+  styleUrls: ['./login-form.component.scss'],
+})
+export class LoginFormRegisterComponent {
+  constructor(public dialogRef: MatDialogRef<LoginFormRegisterComponent>,
+              private userService: UserService,
+              public snackBar: MatSnackBar,
+              @Inject(MAT_DIALOG_DATA) public data: any) {
+  }
+
+  usernameRegisterText: string;
+  passwordRegisterText: string;
+  emailText: string;
+  hideInvalidText = true;
+  errorMessage: string;
+
+  usernameFormControl = new FormControl('', [
+    Validators.required,
+  ]);
+
+  passwordFormControl = new FormControl('', [
+    Validators.required,
+  ]);
+
+  registerValidate() {
+      // Validate user registration by calling signup from UserService
+    if (!this.emailText) {
```

Comment:
You can move the handling for an empty email within the user service signup method to reduce code duplication.

Reply:
That will move the code-duplication to user-service though right?

Is it better to structure it that way or am I missing something that would allow me to just call it once?

If i remove it from registervalidate() from my understanding I'd be doing something similar in UserService like this: https://puu.sh/zYjoX/27fdc5b0c5.png

Reply:
You can do it like this:

```javascript
signup(username: string, password: string, email?: string): Promise<User> {
    const body = {username, password};
    if (email) body.email = email;

    return this.api.post(Version.v1, 'users', body).then((user: User) => {
      this.user = user;
      this.loggedIn = true;
      return user;
    });
  }
```

Reply:
So playing around with it.

In this example, I'm getting an error because body doesn't contain an email type after declaring it as {username, password}. My first thought was to try and add the email element to body after checking if email exists, but I wasn't able to get that to work.

I was thinking perhaps you could define body after you've checked for email but that also runs into its own issues that would result in code duplication. 

I'm likely missing something simple, but how would you approach updating body to have an additional element? I'd imagine it has something to do with updating the object to have 3-fields instead of 2.

Reply:
With the above example, try `const body: {} = {username, password};`

Reply:
Unfortunately I'm still getting a similar type of error doing that. "Property 'email.' does not exist on type '{}'." Here's the code snippet I have for now: https://puu.sh/zYo92/12a5760107.png

I looked a little more into object fields but still a dead end. It's just coming down to being able to add email to body when email exists.

Reply:
I tested it out and while that message does appear, it compiles fine.

Reply:
Oh... so it does,.. Oops... 

I've had that error before where it'd give me problems when i tested it so I just assumed it'd do the same here like a dummy.

Well then.